src/proof/subtype_elim_proof_converter.cpp

```diff
@@ -0,0 +1,317 @@
+    Trace("pf-subtype-elim") << "Premises: " << children << std::endl;
+    Trace("pf-subtype-elim") << "Args: " << cargs << std::endl;
+    AlwaysAssert(false) << "Failed to convert subtyping " << id;
+    return newRes;
+  }
+  // otherwise, newRes is what is proven from the rule without changes,
+  // and resc is what we need to prove.
+  Trace("pf-subtype-elim") << "Introduction of subtyping via rule " << id
+                           << std::endl;
+  Trace("pf-subtype-elim") << "Premises: " << children << std::endl;
+  Trace("pf-subtype-elim") << "Args: " << cargs << std::endl;
+  Trace("pf-subtype-elim") << "...gives " << newRes << std::endl;
+  Trace("pf-subtype-elim") << "...wants " << resc << std::endl;
+  bool success = false;
+  switch (id)
+  {
+    case ProofRule::CONG:
+    case ProofRule::NARY_CONG:
+    {
+      success = true;
+      Node lhs = resc[0];
+      Node rhs = resc[1];
+      std::vector<Node> newChildren;
+      for (size_t i = 0, nchild = lhs.getNumChildren(); i < nchild; i++)
+      {
+        // eqOld is what was proven with the converted children
+        Node eqOld = newRes[0][i].eqNode(newRes[1][i]);
+        // eqNew is what is necessary to prove
+        Node eqNew = lhs[i].eqNode(rhs[i]);
+        // eqOld and eqNew should essentially vary only in their types, e.g.
+        // (= t s) vs. (= (to_real t) (to_real s)),
+        // (= t 0) vs. (= (to_real t) 0.0), etc. We call prove to prove the
+        // updated equality.
+        if (!prove(eqOld, eqNew, cdp))
+        {
+          success = false;
+          break;
+        }
+        newChildren.push_back(eqNew);
+      }
+      // proof with the original rule and updated children should now work
+      if (success)
+      {
+        success = tryWith(id, newChildren, cargs, resc, newRes, cdp);
+      }
+    }
+    break;
+    case ProofRule::ARITH_SUM_UB:
+    {
+      success = true;
+      NodeManager* nm = NodeManager::currentNM();
+      Assert(resc.getNumChildren() == 2);
+      Assert(resc[0].getNumChildren() == children.size());
+      Assert(resc[1].getNumChildren() == children.size());
+      std::vector<Node> newChildren;
+      // reprove what is necessary for the sum for each child
+      for (size_t i = 0, nchild = children.size(); i < nchild; i++)
+      {
+        Node newRel = nm->mkNode(children[i].getKind(), resc[0][i], resc[1][i]);
+        if (!prove(children[i], newRel, cdp))
+        {
+          success = false;
+          break;
+        }
+        newChildren.push_back(newRel);
+      }
+      // proof with the original rule and updated children should now work
+      if (success)
+      {
+        success = tryWith(
+            ProofRule::ARITH_SUM_UB, newChildren, {}, resc, newRes, cdp);
+      }
+    }
+    break;
+    case ProofRule::ARITH_MULT_POS:
+    case ProofRule::ARITH_MULT_NEG:
+    {
+      // This handles the case where we multiply an integer relation by
+      // a rational.
```

Comment:
to make things a bit more self contained, say what is the shape of resc that allows you to build the nodes below